### PR TITLE
Allow validators to register address to send report transactions - Implement keepers

### LINKED ISSUE #848: Allow validators to register address to send report transactions - Implement handlers and keepers

### PR DESCRIPTION
**Pull request by Benzbeeb**

### CODE REVIEW

File: chain/x/zoracle/internal/keeper/reporter.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+func (k Keeper) CheckReporter(ctx sdk.Context, validatorAddress sdk.ValAddress, reporterAddress sdk.AccAddress) bool {
+	store := ctx.KVStore(k.storeKey)
+	return store.Has(types.ReporterStoreKey(validatorAddress, reporterAddress))
+}
+
+func (k Keeper) AddReporter(
+	ctx sdk.Context, validatorAddress sdk.ValAddress, reporterAddress sdk.AccAddress,
+) {
+	store := ctx.KVStore(k.storeKey)
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/reporter.go:24]: Should error if the guy is already a reporter

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/internal/keeper/reporter.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	return store.Has(types.ReporterStoreKey(validatorAddress, reporterAddress))
+}
+
+func (k Keeper) AddReporter(
+	ctx sdk.Context, validatorAddress sdk.ValAddress, reporterAddress sdk.AccAddress,
+) sdk.Error {
+	if k.CheckReporter(ctx, validatorAddress, reporterAddress) {
+		return types.ErrInvalidState(
+			"AddReporter:  (%s) is already a reporter of (%s).",
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/reporter.go]: Remove the extra space. 
Also use ErrDuplicateItem

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/internal/keeper/reporter.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+	return nil
+}
+
+func (k Keeper) RemoveReporter(
+	ctx sdk.Context, validatorAddress sdk.ValAddress, reporterAddress sdk.AccAddress,
+) sdk.Error {
+	if !k.CheckReporter(ctx, validatorAddress, reporterAddress) {
+		return types.ErrItemNotFound(
+			"RemoveReporter: Item notfound for validator address (%s) reporter address (%s).",
```
Comment (sorawit) [inline chain/x/zoracle/internal/keeper/reporter.go]: ... is not a reporter of ...

**Review by sorawit — CHANGES_REQUESTED** [review]

**Comment by sorawit** [comment]
@Benzbeeb please change this PR name as it does not touch handler at all.

File: chain/x/zoracle/internal/types/key.go [hunk 4]
```diff
@@ -110,3 +112,9 @@ func GetValidatorAddressAndExternalID(
 	externalID := ExternalID(binary.BigEndian.Uint64(externalIDBytes))
 	return key[prefixLength+16:], externalID
 }
+
+func ReporterStoreKey(validatorAddress sdk.ValAddress, reporterAddress sdk.AccAddress) []byte {
```
Comment (sorawit) [inline chain/x/zoracle/internal/types/key.go]: Move this function to the location it belongs to.
Add comment
`ReporterStoreKey is a function to generate key for each validator-reporter pair in store.`

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/zoracle/internal/keeper/reporter.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+	return nil
+}
+
+func (k Keeper) RemoveReporter(
+	ctx sdk.Context, validatorAddress sdk.ValAddress, reporterAddress sdk.AccAddress,
+) sdk.Error {
+	if !k.CheckReporter(ctx, validatorAddress, reporterAddress) {
+		return types.ErrItemNotFound(
+			"RemoveReporter: Item notfound for validator address (%s) is not a reporter of reporter address (%s).",
```
Comment (taobun) [inline chain/x/zoracle/internal/keeper/reporter.go]: not found

Comment (taobun) [reply chain/x/zoracle/internal/keeper/reporter.go]: Item not found. %s (reporter addresss) is not a reporter of %s (validator address)

File: chain/x/zoracle/internal/types/key.go [hunk 6]
```diff
@@ -110,3 +112,10 @@ func GetValidatorAddressAndExternalID(
 	externalID := ExternalID(binary.BigEndian.Uint64(externalIDBytes))
 	return key[prefixLength+16:], externalID
 }
+
+// ReporterStoreKey is a function to generate key for each validator-reporter pair in store.
+func ReporterStoreKey(validatorAddress sdk.ValAddress, reporterAddress sdk.AccAddress) []byte {
```
Comment (taobun) [inline chain/x/zoracle/internal/types/key.go]: Move this function after OracleScriptStoreKey

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**